Paper Planes

Mercenaryblue opened the pull request:
Add basic paper planes to the game.
Crafted with paper, now with alt-click crafting.
Thanks to coderbus with the help.

Added PKPenguin321's eye scream damage code. 

Notes and stamps enabled. Just write them up after folding the paper plane.
:cl:
rscadd: Adds a basic paper plane to the game. Fold paper with alt-clicking.
/:cl:

Mercenaryblue committed:
Paper Planes

Add Paper Planes to the game.

Mercenaryblue committed:
Paper planes

Adds basic paper planes to the game.
Crafted with paper on tables.

Mercenaryblue committed:
Merge pull request #1 from Mercenaryblue/paperplane

Paper Planes

RandomMarine commented:
Why tablecrafting? Why not change attack_self to fold it up, considering all it does is call examinate, which is already done by simply shift-clicking.

Bawhoppen commented:
yea this is one of those things that doesn't need table crafting,
tablecrafting makes sense for more complex things but you can easily make a paper airplane by hand

PKPenguin321 commented:
tablecrafting is fine for now

Cheridan commented:
> Information won't transfer between states.

lame

optimumtact commented:
Table crafting is the right way to do this, please ignore the earlier uninformed comments.

Neat feature and welcome back :+1:

Mercenaryblue commented:
Feel free to change the crafting method later yourself, but right now this one works.

Also, good luck folding paper right without using a table IRL.

Lumipharon commented:
Tablecrafting please go.
I can just as eaily fold a plane using a wall, the floor, my knee, someone's ass, a mech, basically anything solid.

Mercenaryblue commented:
Fine. I'll try to figure it out tomorrow.

PKPenguin321 commented:
don't feel obliged mercenaryblue, we already have paper sack crafting that uses tablecrafting, it works fine. those guys are just weenies.

[tablecrafting could use some improvements](https://github.com/tgstation/-tg-station/issues/16107), but that's not your fault and you shouldn't feel obliged to fix it.

Lumipharon commented:
Yeah, not your fault tablecrafting is terrible, but that doesn't change the fact it's still terrible.

PKPenguin321 commented on code/modules/paperwork/paper.dm:
```diff
@@ -404,3 +408,27 @@
+	update_icon()
+
+/obj/item/weapon/paper/paperplane/throw_at(atom/target, range, speed, mob/thrower, spin=0)
+	if(!..())
```
if(..() || !ishuman(target))//if the plane is caught or it hits a nonhuman
     return
    var/mob/living/carbon/human/H = target
    if(prob(15))
      if((H.head && H.head.flags_cover & HEADCOVERSEYES) || (H.wear_mask && H.wear_mask.flags_cover & MASKCOVERSEYES) || (H.glasses && H.glasses.flags_cover & GLASSESCOVERSEYES))
       return
      visible_message("<span class='danger'>\The [src] hits [H] in the eye!</span>")
      H.adjust_blurriness(6)
      H.adjust_eye_damage(rand(6,8))
      H.Weaken(2)
      H.emote("scream")

xxalpha replied:
This is a missing vital part of this feature.

Supermichael777 commented:
just ad a var, make it unreadable when plane, and have its attack self chanfe the name/sprite and set it to folded. jusdicious use of = and = initial.whatever can varedit anything rather then this awkward system.

xxalpha commented on code/modules/paperwork/paper.dm:
```diff
@@ -40,10 +40,14 @@
+		if(!istype(src, /obj/item/weapon/paper/paperplane))  //don't update paperplanes
+			icon_state = "paper_words"
+		return
+	if(istype(src, /obj/item/weapon/paper/paperplane))  //don't update paperplanes to paper
```
Just override the proc
``/obj/item/weapon/paper/paperplane/update_icon()``
``  return ``

xxalpha commented on code/modules/paperwork/paper.dm:
```diff
@@ -40,10 +40,14 @@
 		return
 	if(info)
-		icon_state = "paper_words"
+		if(!istype(src, /obj/item/weapon/paper/paperplane))  //don't update paperplanes
```
Just override the proc

Mercenaryblue replied:
Won't doing so prevent the normal papers from receiving the written sprite?

xxalpha replied:
No

xxalpha replied:
So /obj/item/weapon/paper/paperplane tries to call /obj/item/weapon/paper/paperplane/update_icon(), but since it isn't defined, /obj/item/weapon/paper/update_icon() is called instead.

If you redefine that proc at the /obj/item/weapon/paper/paperplane level, that code is called instead. By doing this and doing nothing in it, you skip the update_icon() code that is meant for normal paper.

JJRcop replied:
basically just remove all this and down in the paperplane section put
```DM
/obj/item/weapon/paper/paperplane/update_icon()
```

JJRcop replied:
No really, listen to @xxalpha, don't bother with the if stuff at the paper level. We call that snowflake, and we don't like it.

xxalpha commented:
:-1: For tablecrafting, :-1: for not saving the text written on the paper

xxalpha commented:
What problems did you have with alt+click? You can ask for help in coderbus.

PKPenguin321 commented:
Why are you guys giving this guy so much shit for tablecrafting when paper bags were added with the same system and had universal high praise

xxalpha commented:
Because tablecrafting a shit

Mercenaryblue commented:
The good news is, I've figured out my mistake and will soon get alt-click paper folding up.

Mercenaryblue committed:
Alt-Click Crafting

Changed to a much better way to craft the paper plane.
Just alt-click your worries away.

Mercenaryblue committed:
Merge pull request #2 from Mercenaryblue/paperplane

Paperplane

Mercenaryblue committed:
PKPenguin321's Eye Scream! Yum!

Added PKPenguin321 's code for random eye damage.

Mercenaryblue committed:
Merge pull request #3 from Mercenaryblue/paperplane

Alt-Crafting and Random Eye Scream damage

Mercenaryblue commented:
Alt-Click Crafting implemented
Eye Scream Damage implemented
Thanks for the feedback. I was just too tired last night.

PKPenguin321 commented on code/modules/paperwork/paper.dm:
```diff
@@ -401,6 +405,53 @@
+		H.adjust_eye_damage(rand(6,8))
+		H.Weaken(2)
+		H.emote("scream")
+		H.say("My eyes!!")
```
Forced says are generally frowned upon, I personally don't care but you should tread carefully

RemieRichards replied:
Remove it.

Mercenaryblue committed:
Fair enough

No more screaming "My eyes!" automatically

Mercenaryblue committed:
Merge pull request #4 from Mercenaryblue/paperplane

Fair enough

PKPenguin321 commented on code/modules/paperwork/paper.dm:
```diff
@@ -401,6 +405,52 @@
+	update_icon()
+
+/obj/item/weapon/paper/paperplane/throw_at(atom/target, range, speed, mob/thrower, spin=0)
+	if(!..() || !ishuman(target))//if the plane is caught or it hits a nonhuman
```
I believe you want ..(), not !..()

From testing with things such as bolas I've found that ..() has the intended effect of returning if the item is caught. The inverse would likely just return if the item isn't caught.

Mercenaryblue replied:
Won't leaving it at if..() means that if it loads, it'll return at 447, thus cancelling the rest of the code?

PKPenguin321 replied:
you'd think that but no
i'd recommend ingame testing but i assure you i'm right on this, it's used in things like bolas and bluespace tomatoes and both work correctly

Mercenaryblue replied:
Allright I'll test it out tomorrow

Mercenaryblue replied:
From the test:
**With if(!..()**
-Target caught paperplane
-Eye Scream Procs occured without trouble

***With if(..())**
-Target Caught Paperplane
-No Eye Scream Procs (35 throws)

I'll deduce the if(!..() was the correct code to put with return, since it returned the best results.

phil235 replied:
You're using the wrong proc. You need `/obj/item/weapon/paper/paperplane/throw_impact(atom/hit_atom)`.

PKPenguin321 replied:
oops yeah what phil said

Commit:
Paper planes now keep their old text

JJRcop commented:
Mercenaryblue#5

I didn't fix any of the issues, because there is learning to be had.
I just added
>information transfer between states

Mercenaryblue committed:
Merge pull request #5 from JJRcop/mumblegrumble

Paper planes now keep their old text

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into paperplane

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master' into paperplane

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into paperplane

Mercenaryblue committed:
Doesn't fuck with talismans 

Prevent folding talismans and breaking the game mode.

Mercenaryblue committed:
Doesn't fuck with cult

Prevent folding talismans into powerless planes

Mercenaryblue committed:
Merge pull request #6 from Mercenaryblue/paperplane

Doesn't fuck with cult

Mercenaryblue committed:
Fix

Mercenaryblue committed:
Merge remote-tracking branch 'origin/paperplane' into paperplane

# Conflicts:
#	code/modules/paperwork/paper.dm

PKPenguin321 committed:
corrects the proc for plane throwing

switches it to throw_impact
restores throw_at proc override

PKPenguin321 committed:
fixes travis

my bad

Mercenaryblue committed:
paperplane update_icon() fixes

Mercenaryblue committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	code/modules/paperwork/paper.dm

PKPenguin321 committed:
!..() -> ..()

forgot to do this when i switched the proc

Mercenaryblue committed:
Revert Throw & Others

Revert Throw code to previous version (working fine)
Update_Icon() improved
Added burning paperplane overlays.
Sprite in next commit.

Mercenaryblue committed:
Revert Throw, Add Burning

Reverted to old throw code
Added burning code
Redone update_icon()

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/tgstation/master' into paperplane

Mercenaryblue committed:
Merge remote-tracking branch 'refs/remotes/origin/master' into paperplane

# Conflicts:
#	code/modules/paperwork/paper.dm

Mercenaryblue committed:
Throw Code Re-Implemented

Everything's fixed now. Right?!

Mercenaryblue committed:
Merge pull request #8 from Mercenaryblue/paperplane

Throw Code & update_icon()

Mercenaryblue commented:
Throw code fixed.
Update_Icon() code improved
Burning Plane code implemented

That's it. We got them all. Quick Alt-Click Crafting, Anti-spin throw, Info transfer, Stamps, Icon Updates, Anti-Cult checks, Eye Damage proc, etc... This feature is done... for now.

Chocobro commented:
are the sprites used my ones from ages ago that i posted on the forums? on phone, can't into dmi

Mercenaryblue commented:
Nah, I've made the sprites myself.

Chocobro commented:
alrighty then. merge soon fam

JJRcop commented on code/modules/paperwork/paper.dm:
```diff
@@ -404,3 +404,88 @@
+/obj/item/weapon/paper/paperplane/fire_act()
+	..(0)
+	icon_state = "paperplane"
+	info = "[stars(info)]"
```
you can nix this line, already covered in paper's fire_act

Mercenaryblue replied:
Yeah... the paper's fire_act is why I coded this in.

Without the code the game would update the paper back to his original form before setting it on fire. The code allows the paperper plane to keep its direction, sprite, and adds in the flame overlay.

Just my experience.

JJRcop commented on code/modules/paperwork/paper.dm:
```diff
@@ -404,3 +404,88 @@
+	info = "[stars(info)]"
+	update_icon()
+
+/obj/item/weapon/paper/paperplane/extinguish()
```
you can completely nix this

JJRcop commented on code/modules/paperwork/paper.dm:
```diff
@@ -404,3 +404,88 @@
+	
+/obj/item/weapon/paper/paperplane/update_icon()
+	if(burn_state == ON_FIRE)
+		overlays += "paperplane_onfire"
```
```DM
icon_state = "paperplane_onfire"
```

And you need to create a paperplane_onfire.

Mercenaryblue replied:
Fucking hell did I forget to add them to the commit

Mercenaryblue committed:
Forgot sprites

Adding sprites

Mercenaryblue committed:
Updating paper

Mercenaryblue committed:
Merge remote-tracking branch 'origin/master'

# Conflicts:
#	code/modules/paperwork/paper.dm

Mercenaryblue committed:
Paperplane Final Code

Mercenaryblue committed:
Merge pull request #9 from Mercenaryblue/paperplane

Forgot sprites

xxalpha commented on code/modules/paperwork/paper.dm:
```diff
@@ -404,3 +404,117 @@
+
+=======
+		
+>>>>>>> origin/master
```
you fucked up

phil235 commented:
@Mercenaryblue  You shouldn't use your master directly. Make a branch from your master and make your PR with that branch.
